src/rmutil/args.c

```diff
@@ -0,0 +1,125 @@
+#include "args.h"
+#include "redismodule.h"
+#include <float.h>
+#include <math.h>
+#include <string.h>
+
+int AC_Advance(ArgsCursor *ac) {
+  if (ac->offset == ac->argc) {
+    return AC_ERR_NOARG;
+  } else {
+    ac->offset++;
+  }
+  return 0;
+}
+
+#define MAYBE_ADVANCE()            \
+  if (!(flags & AC_F_NOADVANCE)) { \
+    AC_Advance(ac);                \
+  }
+
+static int tryReadAsDouble(ArgsCursor *ac, long long *ll, int flags) {
+  double dTmp = 0.0;
+  if (AC_GetDouble(ac, &dTmp, flags | AC_F_NOADVANCE)) {
+    return AC_ERR_PARSE;
+  }
+  if (flags & AC_F_COALESCE) {
+    *ll = dTmp;
+    return 0;
+  }
+
+  if ((double)(long long)dTmp != dTmp) {
+    return AC_ERR_PARSE;
+  } else {
+    *ll = dTmp;
+    return 0;
+  }
+}
+
+int AC_GetLongLong(ArgsCursor *ac, long long *ll, int flags) {
+  if (ac->offset == ac->argc) {
+    return AC_ERR_NOARG;
+  }
+  if (ac->type == AC_TYPE_RSTRING) {
+    if (RedisModule_StringToLongLong(AC_CURRENT(ac), ll) == REDISMODULE_ERR &&
+        tryReadAsDouble(ac, ll, flags)) {
+    }
+  } else {
+    char *endptr = AC_CURRENT(ac);
+    *ll = strtoll(AC_CURRENT(ac), &endptr, 10);
+    if (*endptr != '\0' || *ll == LLONG_MIN || *ll == LLONG_MAX) {
+      if (tryReadAsDouble(ac, ll, flags) != 0) {
+        return AC_ERR_PARSE;
+      }
+    }
+  }
+  if ((flags & AC_F_GE0) && *ll < 0) {
+    return AC_ERR_ELIMIT;
+  }
+  // Do validation
+  if ((flags & AC_F_GE1) && *ll < 1) {
+    return AC_ERR_ELIMIT;
+  }
+  MAYBE_ADVANCE();
+  return 0;
```

Comment:
I think its better to return AC_OK and not 0, though its the same its less confusing because most times zero means failure and here zero is success.

Reply:
Usually, when the return options are binary, (sucess/fail) then 0 means failure and ok means success. In POSIX and in most other 'robust' systems, 0 means ok, and a nonzero return means some kind of error. I will change these to mean AC_OK, but AC_OK will still be 0.

Reply:
Usually, when the return options are binary, (sucess/fail) then 0 means failure and ok means success. In POSIX and in most other 'robust' systems, 0 means ok, and a nonzero return means some kind of error. I will change these to mean AC_OK, but AC_OK will still be 0.

Reply:
This is what I meant